File: bytom/recommend.php
<?php
	if(!isset($_GET['id'])){
		die();
	}

	#调用推荐系统
	$id = $_GET['id'];
	#下面的 2>&1表示输出错误信息
	$id = urlencode($id);
    $program="PYTHONIOENCODING=utf-8 python3 ./python/lib/recommend.py {$id} 2>&1";
    exec($program);

	$redis = new Redis();
	$redis->connect("localhost",6379);
	$res = $redis->hGet('recommend',$_GET['id']);


	//获取推荐结果
	$html = ""; 
	$mysql = new Mysqli("localhost","root","fang","bytom");
	$mysql->query("set names utf8");
	if($res){
		//已经有结果
		$id_list = eval("return $res;");
		
		$f_id = $id_list[0];
		$sql = "select id,title,info,url from article where id=$f_id";
		$data = $mysql->query($sql);
		$data = $data->fetch_all(MYSQLI_ASSOC);
		$html = $html."<h2><span class='red'>【站长推荐】</span> 
          <a target='_blank' href='".$data[0]['url']."' onclick='recommend(".$data[0]['id'].")'>".$data[0]['title']."</a></h2><p class='note' style='margin-left:10px'><a target='_blank' href='".$data[0]['url']."'>".$data[0]['info']."</a></p><div class='relatescom'><ul>";
		for($i=1;$i<sizeof($id_list);$i++){
			$id = $id_list[$i];
			$sql = "select id,title,info,url from article where id=$id";
			$data = $mysql->query($sql);
			$data = $data->fetch_all(MYSQLI_ASSOC);
			$html = $html."<li><a target='_blank' href='".$data[0]['url']."' onclick='recommend(".$data[0]['id'].")'>".$data[0]['title']."</a></li>";
		}
		$html = $html."</ul></div>";
	}
	echo $html;

File: bytom/python/lib/recommend.py
#coding=utf-8
import jieba, re
import pymysql 
import os
import sys
import redis
from urllib import parse
from gensim import corpora,models,similarities

#url解码
a_id = re.escape(sys.argv[1])

#查询所有文本数据
db = pymysql.connect("localhost","root","fang","bytom")
cursor = db.cursor()
sql1 = "set names utf8"
sql = "select id,content,title from content_text"

cursor.execute(sql1)
cursor.execute(sql)
res = cursor.fetchall()
id_list = []
text_list = []
test = []
for data in res:
	#得到分词矩阵
    if(data[0]!=int(a_id)):
    	id_list.append(data[0])
    	text_list.append(data[1].split(" "))
    else:
    	test = data[1].split(" ")
    	


#制作词库
dictionary = corpora.Dictionary(text_list)
corpus = [dictionary.doc2bow(doc) for doc in text_list]
#测试集
test = dictionary.doc2bow(test)    

#训练模型
tfidf = models.TfidfModel(corpus)

index = similarities.SparseMatrixSimilarity(tfidf[corpus], num_features=len(dictionary.keys()))
sim = index[tfidf[test]]
#获取相似度结果
res = sorted(enumerate(sim), key=lambda item: -item[1])

res = res[:5]

#存储推荐结果到redis
res_list = []
for i in res:
	res_list.append(id_list[i[0]])
print(res_list)
pool = redis.ConnectionPool(host='localhost', port=6379, decode_responses=True)  
r = redis.Redis(connection_pool=pool)
r.hset("recommend",a_id,str(res_list))


   
